core/client/fs/src/main/java/alluxio/client/block/stream/SharedGrpcDataReader.java

```diff
@@ -0,0 +1,190 @@
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A shared gRPC data reader that cache blocks data for multi-thread accessing.
+ *
+ * It follows GrpcDataReader protocol and takes strong assumption:
+ * Parallel read to the same file happens on the same time, so that read request is
+ * serialized by kernel
+ */
+@NotThreadSafe
+public class SharedGrpcDataReader implements DataReader {
+  private static final Logger LOG = LoggerFactory.getLogger(SharedGrpcDataReader.class);
+  private static final int BLOCK_LOCK_NUM = 32;
+  private static final ReentrantReadWriteLock[] BLOCK_LOCKS;
+  /** A map from block id to the block's cached data reader. */
+  @GuardedBy("BLOCK_LOCKS")
+  private static final ConcurrentHashMap<Long, BufferCachingGrpcDataReader> BLOCK_READERS
+      = new ConcurrentHashMap<>();
+
+  static {
+    BLOCK_LOCKS = new ReentrantReadWriteLock[BLOCK_LOCK_NUM];
+    for (int i = 0; i < BLOCK_LOCK_NUM; i++) {
+      BLOCK_LOCKS[i] = new ReentrantReadWriteLock();
+    }
+  }
+
+  private static ReentrantReadWriteLock getBlockLock(long blockId) {
+    return BLOCK_LOCKS[(int) (blockId % BLOCK_LOCKS.length)];
+  }
+
+  private final long mBlockId;
+  private final BufferCachingGrpcDataReader mCachedDataReader;
+  private final long mChunkSize;
+
+  /** The next pos to read. */
+  private long mPosToRead;
+
+  /**
+   * Creates an instance of {@link SharedGrpcDataReader}.
+   *
+   * @param readRequest the read request
+   * @param reader the cached Grpc data reader for the given block
+   */
+  private SharedGrpcDataReader(ReadRequest readRequest, BufferCachingGrpcDataReader reader) {
+    mChunkSize = readRequest.getChunkSize();
+    mPosToRead = readRequest.getOffset();
+    mBlockId = readRequest.getBlockId();
+    mCachedDataReader = reader;
+  }
+
+  @Override
+  public long pos() {
+    return mPosToRead;
+  }
+
+  /**
+   * Seeks to a specific position.
+   *
+   * @param pos the position to seek to
+   */
+  public void seek(long pos) {
+    mPosToRead = pos;
+  }
+
+  @Override
+  @Nullable
+  public DataBuffer readChunk() throws IOException {
+    int index = (int) (mPosToRead / mChunkSize);
+    DataBuffer chunk = mCachedDataReader.readChunk(index);
+    if (chunk == null) {
+      return null;
+    }
+    ByteBuffer bb = chunk.getReadOnlyByteBuffer();
+    // Force to align to chunk size
+    bb.position((int) (mPosToRead % mChunkSize));
+    mPosToRead += mChunkSize - mPosToRead % mChunkSize;
+
+    return new NioDataBuffer(bb, bb.remaining());
+  }
+
+  @Override
+  @Nullable
+  public DataBuffer readChunkIfReady() throws IOException {
+    // I'm naive, I'm reading chunks anyway
```

Comment:
I'm not sure what this means. Does this mean this reader is always going to be ready?

Reply:
removed